### PR TITLE
*: fix schema change validation for partitioned table when transaction commit

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

If the schema changes during a transaction, the transaction should check whether it's snapshot schema is still valid.

For example:

```
begin;  // s1
insert into t values ... // s1
alter table t add index idx(id) // s2
commit; // s1 should fai because the schema has changed 
```

### What is changed and how it works?

In the transaction, we collect the affected table IDs.
During the schema change, we collect the changed table IDs' history for each version.
And we compare the transaction modified table IDs with the schema change history when the transaction commits, and reject the operation if those IDs overlap.

This bug happens on the partitioned tables because we use `physicalTableID` in the transaction's modification and use `tableID` in the schema change history.

I fix it by recording the `physicalTableID` in the schema change history.

### Check List 

Tests 

 - Unit test

The original code is not well tested for all cases.
We should improve test coverage here.

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @zimulala @jackysp @coocood

**Comment by jackysp** [comment]
@bb7133 @crazycs520 PTAL

File: infoschema/builder.go [hunk 1]
```diff
@@ -89,21 +87,37 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) ([]int64, erro
...
 	return tblIDs, nil
 }
 
+func appendAffectedIDs(affected []int64, tblInfo *model.TableInfo) []int64 {
+	if pi := tblInfo.GetPartitionInfo(); pi != nil {
+		for _, def := range pi.Definitions {
+			affected = append(affected, def.ID)
+		}
+	}
+	affected = append(affected, tblInfo.ID)
```
Comment (crazycs520) [inline infoschema/builder.go]: Please add `select for update` test for this. Prevent someone deleting this line by mistake.

Comment (crazycs520) [reply infoschema/builder.go]: Look like it's redundancy, forgot this.

**Comment by tiancaiamao** [comment]
PTAL @zimulala

**Comment by tiancaiamao** [comment]
PTAL @zimulala

File: domain/domain.go [hunk 2]
```diff
@@ -354,10 +354,10 @@ func (do *Domain) Reload() error {
 	}
 
 	var (
-		fullLoad        bool
-		changedTableIDs []int64
+		fullLoad                bool
+		changedPhysicalTableIDs []int64
 	)
-	neededSchemaVersion, changedTableIDs, fullLoad, err = do.loadInfoSchema(do.infoHandle, schemaVersion, ver.Ver)
```
Comment (zimulala) [inline domain/domain.go]: We change this variable name, then do we need to change the argument name of `changedTableIDs` in`schemaValidator.Update`?

File: infoschema/builder.go (hunk 1 above)
Comment (zimulala) [inline infoschema/builder.go]: Could we put this line to line111? That feels like it can be checked earlier if the DML is related to recent mode changes

**Comment by tiancaiamao** [comment]
s/relatedTableID/relatedPhysicalTableID  in schema validator.  @zimulala @jackysp @crazycs520

**Pull request merged**